plugins/out_s3/s3.c

```diff
@@ -1105,46 +1094,32 @@ static int upload_data(struct flb_s3 *ctx, struct s3_file *chunk,
 
 put_object:
 
-    /*
-     * remove chunk from buffer list- needed for async http so that the
-     * same chunk won't be sent more than once
-     */
     ret = s3_put_object(ctx, tag, file_first_log_time, body, body_size);
     if (ctx->compression == FLB_AWS_COMPRESS_GZIP) {
         flb_free(payload_buf);
     }
     if (ret < 0) {
-        /* re-add chunk to list */
         if (chunk) {
             chunk->failures += 1;
             if (chunk->failures > ctx->ins->retry_limit){
-                s3_retry_warn(ctx, tag, chunk->input_name, create_time, FLB_FALSE);
-                s3_store_file_delete(ctx, chunk);
```

Comment:
after failures is greater than retry limit why is the chunk not deleted?

Reply:
I see. You're unifying all the chunk maintenance logic to send. This makes things simpler.

Reply:
also we have to remove from the upload_queue list before deleting else segfault IIRC

Reply:
I just realized we need to support the unlimited retries option here: https://docs.fluentbit.io/manual/administration/scheduling-and-retries